src/lib/HasuraGraphiQL.tsx

```diff
@@ -0,0 +1,563 @@
+import * as React from "react";
+import GraphiQL, { GraphiQLProps } from "graphiql";
+import GraphiQLExplorer from "graphiql-explorer";
+import {
+  buildClientSchema,
+  getIntrospectionQuery,
+  GraphQLSchema,
+} from "graphql";
+import CodeExporter from "graphiql-code-exporter";
+import snippets from "./snippets";
+import makeFetcher from "./fetcher";
+import {
+  transformHeaders,
+  untransformHeaders,
+  edited2DArray,
+  toggleCacheDirective,
+  hasCacheDirective,
+} from "./utils";
+import {
+  IconChevronRight,
+  IconChevronDown,
+  IconCross,
+  IconEye,
+  IconInfoCircle,
+  IconCheckCircle,
+} from "./Icons";
+import CustomGraphiQL from "./CustomGraphiQL";
+import Spinner from "./Spinner";
+
+import "graphiql/graphiql.css";
+import "graphiql-code-exporter/CodeExporter.css";
+import "./styles.css";
+
+export default function HasuraGraphiQL({
+  defaultUrl = "",
+  defaultHeaders = {},
+  defaultQuery = "",
+  isCloud = false,
+  defaultVariables = "",
+  hiddenHeaders = ["x-hasura-admin-secret"],
+  graphiQLOptions = {},
+  explorerOptions = {},
+  customToolbar = null,
+}: {
+  defaultUrl?: string;
+  defaultHeaders?: Record<string, string>;
+  defaultQuery?: string;
+  isCloud?: boolean;
+  defaultVariables?: string;
+  hiddenHeaders?: string[];
+  graphiQLOptions?: Omit<GraphiQLProps, "fetcher">;
+  explorerOptions?: Record<string, any>;
+  customToolbar?: React.ReactNode;
+}) {
+  const [loading, setLoading] = React.useState(true);
+  const [schema, setSchema] = React.useState<GraphQLSchema | undefined>(
+    undefined
+  );
+  const [url, setUrl] = React.useState(defaultUrl);
+  const [urlInput, setUrlInput] = React.useState(defaultUrl);
+  const [query, setQuery] = React.useState<string | undefined>(defaultQuery);
+  const [headers, setHeaders] = React.useState(
+    untransformHeaders(defaultHeaders, hiddenHeaders)
+  );
+  const [headersInput, setHeadersInput] = React.useState(
+    untransformHeaders(defaultHeaders, hiddenHeaders)
+  );
+  const [urlCollapsed, setUrlCollapsed] = React.useState(false);
+  const [headersCollapsed, setHeadersCollapsed] = React.useState(false);
+  const [relay, setRelay] = React.useState(false);
+  const [codeExporterVisible, setCodeExporterVisible] = React.useState(false);
+  const [explorerVisible, setExplorerVisible] = React.useState(true);
+  const [errorShown, setErrorShown] = React.useState(false);
+  const [error, setError] = React.useState(null);
+  const [responseTime, setResponseTime] = React.useState<number | null>(null);
+  const [responseSize, setResponseSize] = React.useState(0);
+  const [explorerWidth, setExplorerWidth] = React.useState(300);
+  const [resizing, setResizing] = React.useState(false);
+  const [isCached, setIsCached] = React.useState(false);
```

Comment:
Before I jump in and implement useReducer, let me pick your brains for some time.

With useReducer, the state would look something like this - 

```javascript
const [{
      loading, 
      schema, 
      query, 
      headers, 
      headersInput, 
      urlCollapsed, 
      headersCollapsed, 
      relay, 
      codeExporterVisible, 
      explorerVisible, 
      errorVisible, 
      responseTime, 
      responseSize, 
      isCached, 
      isResizing, 
      explorerWidth}, 
    dispatch ] = React.useReducer(someReducer, {
      loading:false, 
      schema:null, 
      query:"", 
      headers:..., 
      headersInput:.., 
      urlCollapsed:false, 
      headersCollapsed:false, 
      relay:false, 
      codeExporterVisible:false, 
      explorerVisible:true, 
      errorVisible:false, 
      responseTime:null, 
      responseSize:"", 
      isCached:false, 
      isResizing:false, 
      explorerWidth:300})
```
This doesn't seem to make things any better. If you were thinking of using useReducer in some other way, do let me know.

Reply:
Well, for me, there is 2 separate concern I have here : 
1. Some variable are updated together (eg update headers) and should be a reducer to be sure that it's consistent
2. The number of state / variable in a reducer shows that the component is too big / need splitting up

If we can split it up, it would be great, but if not, a reducer for all of the things that are updated in batch (eg update headers) is a safer approch

Reply:
@beaussan - Yes. Splitting the component will help organise the state and the code better. I judged that that since the component is only ~500 lines of code and the sub-components cannot be reused anyway (since it's a lib), I could just make everything encapsulated in one function. Guess my judgement was wrong since both of you feel the code is hard to read/maintain. I'll refactor and split up the component in smaller parts.

About updating headers, if you don't mind, could you please elaborate how using a reducer is a safer approach? I have a function called edited2DArray to immutably update arrays and I use that to to update the headers array. I thought immutably updating array state was ok. Please correct me if I am wrong.

Reply:
Even if the component is used once, splitting for readability / maintainability is always a good thing.

For the headers, what I mean is, for me, useState !== useReducer, here is a small example of what I mean :

```ts
  const [{}, dispatch] = useReducer((state, { action, payload }) => {
    switch (action) {
      case 'updateHeaders': {
        if (payload === state.headers) {
          return state;
        }
        return {
          ...state,
          schema: undefined,
          loading: true,
          headers: payload,
        };
      }
    default:
        return state;
    }
  }, initialValues);
```
In the reducer, we speak of action that have side effect (like here update header that can update the loading / schema change, and it's the only way to update the headers.

Is that clearer ?

Reply:
Now I understand what you mean. Thanks a lot for the detailed reply!